src/main/java/it/smartcommunitylab/aac/tos/controller/TosController.java

```diff
@@ -0,0 +1,117 @@
+ */
+
+package it.smartcommunitylab.aac.tos.controller;
+
+import java.util.Locale;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.validation.Valid;
+import javax.validation.constraints.Pattern;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.authentication.InsufficientAuthenticationException;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+import it.smartcommunitylab.aac.SystemKeys;
+import it.smartcommunitylab.aac.common.NoSuchProviderException;
+import it.smartcommunitylab.aac.common.NoSuchRealmException;
+import it.smartcommunitylab.aac.common.NoSuchUserException;
+import it.smartcommunitylab.aac.core.AuthenticationHelper;
+import it.smartcommunitylab.aac.core.MyUserManager;
+import it.smartcommunitylab.aac.core.UserDetails;
+import it.smartcommunitylab.aac.core.auth.RealmAwareAuthenticationEntryPoint;
+
+@Controller
+@RequestMapping
+public class TosController {
+
+	private final Logger logger = LoggerFactory.getLogger(getClass());
+
+	@Autowired
+	private AuthenticationHelper authHelper;
+
+	@Autowired
+	private MyUserManager myUserManager;
+
+	@GetMapping("/terms/{approveTOS}")
+	public String terms(@PathVariable @Valid @Pattern(regexp = SystemKeys.SLUG_PATTERN) String approveTOS,
+			HttpServletRequest request, Model model, Locale locale)
+			throws NoSuchProviderException, NoSuchUserException, NoSuchRealmException {
+
+		UserDetails user = authHelper.getUserDetails();
+
+		if (user == null) {
+			throw new InsufficientAuthenticationException("error.unauthenticated_user");
+		}
+
+		String realm = user.getRealm();
+		model.addAttribute("acceptUrl", "/terms/accept");
+		model.addAttribute("realm", realm);
+		model.addAttribute("displayName", realm);
+
+		if (approveTOS.equals("true")) {
+			model.addAttribute("rejectUrl", "/terms/refuse");
+			return "tos/tos_approval";
+		}
+
+		return "tos/tos_ok";
+
+	}
+
+	@PostMapping("/terms/accept")
+	public String termsAccepted(HttpServletRequest request, Model model, Locale locale)
+			throws NoSuchProviderException, NoSuchUserException, NoSuchRealmException {
+
+		UserDetails user = authHelper.getUserDetails();
+
+		if (user == null) {
+			throw new InsufficientAuthenticationException("error.unauthenticated_user");
+		}
+
+		request.getSession().setAttribute("termsManaged", "true");
+		this.logger.debug("terms of service approved");
+
+		return "redirect:/";
+	}
+
+	@GetMapping("/terms/refuse")
+	public String termsRefused(HttpServletRequest request, Model model, Locale locale)
```

Comment:
drop this. The flow is 
`tos form -> tos post handler -> view`

everything else is handled in the filter